hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/SmallFileStreamDataChannel.java

```diff
@@ -0,0 +1,176 @@
+      throws StorageContainerException {
+    super(file, container.getContainerData(), metrics);
+    this.blockManager = blockManager;
+    this.kvContainer = container;
+    this.realLen = (int) dataLen;
+  }
+
+  @Override
+  ContainerProtos.Type getType() {
+    return ContainerProtos.Type.PutSmallFile;
+  }
+
+  @Override
+  public int write(ByteBuffer src) throws IOException {
+    int srcLen = src.capacity();
+
+    if (srcLen == 0) {
+      forceClose = true;
+      return 0;
+    } else if (writeLen + srcLen > realLen) {
+
+      if (metadataLen > 0) {
+        metadataLen += srcLen;
+        metadata.add(src);
+      } else {
+        metadataLen += (writeLen + srcLen - realLen);
+
+        int dataLen = srcLen - metadataLen;
+        byte[] data = new byte[dataLen];
+        src.get(data, 0, dataLen);
+        super.write(ByteBuffer.wrap(data));
+
+        byte[] meta = new byte[metadataLen];
+        src.get(meta, dataLen, metadataLen);
+        metadata.add(ByteBuffer.wrap(meta));
+      }
+    } else {
+      super.write(src);
+    }
+    writeLen += srcLen;
+    return srcLen;
+  }
+
+  private ByteString asByteString() {
+    ByteBuffer buffer = ByteBuffer.allocate(metadataLen);
+    for (ByteBuffer b : metadata) {
+      buffer.put(b);
+    }
+    buffer.flip();
+    return ByteString.copyFrom(buffer);
+  }
+
+  @Override
+  public void close() throws IOException {
+    if (forceClose) {
+      super.close();
+      return;
+    }
+
+    if (writeLen <= realLen || metadataLen <= 0) {
+      String msg = "Put small file write length mismatch realLen: " +
+          realLen + " writeLen: " + writeLen + " metadataLen: " + metadataLen;
+      throw new StorageContainerException(msg,
+          ContainerProtos.Result.PUT_SMALL_FILE_ERROR);
+    }
+
+    ContainerProtos.ContainerCommandRequestProto request =
+        ContainerCommandRequestMessage.toProto(asByteString(), null);
+
+    if (!request.hasPutSmallFile()) {
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Malformed Put Small File request. trace ID: {}",
+            request.getTraceID());
+      }
+      throw new StorageContainerException("Malformed Put Small File request.",
+          ContainerProtos.Result.PUT_SMALL_FILE_ERROR);
+    }
+
+    ContainerProtos.PutSmallFileRequestProto putSmallFileReq =
+        request.getPutSmallFile();
+
+    ContainerProtos.ChunkInfo chunkInfoProto = putSmallFileReq.getChunkInfo();
+    ChunkInfo chunkInfo = ChunkInfo.getFromProtoBuf(chunkInfoProto);
+    Preconditions.checkNotNull(chunkInfo);
+
+    blockData = BlockData.getFromProtoBuf(
+        putSmallFileReq.getBlock().getBlockData());
+    Preconditions.checkNotNull(blockData);
+
+    List<ContainerProtos.ChunkInfo> chunks = new LinkedList<>();
+    chunks.add(chunkInfoProto);
+    blockData.setChunks(chunks);
+
+    super.close();
+  }
+
+  @Override
+  public void link(RaftProtos.LogEntryProto entry) throws IOException {
```

Comment:
@szetszwo This is used to update the BCSID in blockData

Reply:
We should update it using the same way as in ContainerStateMachine.applyTransaction(..).  It first build a context and the submit it for execution; see https://github.com/apache/ozone/blob/e8af5823c6bcddd067062138a55229b0565ee59b/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java#L837-L871